pkg/dnscache/dnscache_test.go

```diff
@@ -0,0 +1,106 @@
+package dnscache_test
+
+import (
+	"bufio"
+	"compress/gzip"
+	"encoding/json"
+	"os"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/aquasecurity/tracee/pkg/dnscache"
+	"github.com/aquasecurity/tracee/pkg/logger"
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+var eventDump []trace.Event = []trace.Event{}
+
+func init() {
+	zippedEventsFile, err := os.OpenFile("./dns_events_test.gz", os.O_RDONLY, os.ModeAppend)
+	if err != nil {
+		panic(err)
+	}
+	gzipReader, err := gzip.NewReader(zippedEventsFile)
+	if err != nil {
+		panic(err)
+	}
+	defer gzipReader.Close()
+
+	scanner := bufio.NewScanner(gzipReader)
+	for scanner.Scan() {
+		event := scanner.Bytes()
+		var e trace.Event
+		err := json.Unmarshal(event, &e)
+		if err != nil {
+			logger.Errorw("Invalid json in " + string(event) + ": " + err.Error())
+		} else {
+			eventDump = append(eventDump, e)
+		}
+	}
+
+}
+
+func TestNodeDnsCache(t *testing.T) {
+	dnsCache, err := dnscache.New(dnscache.Config{
+		Enable:    true,
+		CacheSize: 6000,
+	})
+
+	require.NoError(t, err)
+
+	for i := 0; i < len(eventDump); i++ {
+		evt := eventDump[i]
+		evt.Timestamp = int(time.Now().UnixNano()) // modify the timestamp so expiry/ttl logic won't work
+		err = dnsCache.Add(&evt)
+		require.NoError(t, err)
+	}
+
+	t.Log("length: ", len(eventDump))
+
+	// test from query root down
+	res, err := dnsCache.Get("google.com")
+	require.NoError(t, err)
+	assert.Equal(t, res.DNSResults(), []string{"google.com"})
+	assert.ElementsMatch(t, res.IPResults(), []string{
+		"142.251.111.113",
+		"142.251.111.102",
+		"142.251.111.101",
+		"142.251.111.139",
+		"142.251.111.100",
+		"142.251.111.138"},
+	)
+	res, err = dnsCache.Get("www.cooldictionary.com")
+	require.NoError(t, err)
+	assert.Equal(t, res.DNSResults(), []string{"www.cooldictionary.com"})
+	assert.Equal(t, res.IPResults(), []string{"104.196.52.56"})
+
+	// test from index down
+	res, err = dnsCache.Get("a229.g.akamai.net")
+	require.NoError(t, err)
+	assert.ElementsMatch(t, res.DNSResults(), []string{
+		"wdw2.wdpromedia.com", "wdw2.wdpromedia.com.edgesuite.net", "a229.g.akamai.net",
+	})
+
+	// test from ip leaves up (other ips will not be included)
+	res, err = dnsCache.Get("31.13.66.19")
+	require.NoError(t, err)
+	assert.ElementsMatch(t, res.DNSResults(), []string{
+		"connect.facebook.net", "scontent.xx.fbcdn.net",
+	})
+	assert.Equal(t, res.IPResults(), []string{"31.13.66.19"})
+	res, err = dnsCache.Get("2606:4700:10::6816:1f")
+	require.NoError(t, err)
+	assert.ElementsMatch(t, res.DNSResults(), []string{"dailymaverick.co.za"})
+	assert.Equal(t, res.IPResults(), []string{"2606:4700:10::6816:1f"})
+
+	// test reverse ip query
+	res, err = dnsCache.Get("231.92.80.190.in-addr.arpa")
+	require.NoError(t, err)
+	assert.Equal(t, res.DNSResults(), []string{"231-92-pool.dsl.gol.net.gy"})
+	assert.Equal(t, res.IPResults(), []string{"231.92.80.190"})
+
+	// possibly add more tests?
```

Comment:
I would add a very small LRU cache (so evictions happen very frequently) and stress test paralelization. I have doubts if you won't be able to cause races in the TryLock() call I commented about.

Reply:
Already explained why I don't think the deadlock is likely (actually the TryLock resolved an initial deadlock bug), but no harm in adding, and if it proves me wrong all the better.

Reply:
Ok i've added such a stress test, it helped finding other incidental bugs, but there's no other deadlock.